tests/model_fields/test_choiceenum.py

```diff
@@ -0,0 +1,282 @@
+import datetime
+import decimal
+import enum
+import ipaddress
+import uuid
+
+from django.db import models
+from django.db.models.enums import ChoiceEnumMeta
+from django.test import SimpleTestCase, TestCase
+from django.utils.functional import Promise
+from django.utils.translation import gettext_lazy as _
+
+from .models import Foo
+
+
+class Suit(models.ChoiceIntEnum):
+    DIAMOND = 1, _('Diamond')
+    SPADE = 2, _('Spade')
+    HEART = 3, _('Heart')
+    CLUB = 4, _('Club')
+
+
+class YearInSchool(models.ChoiceStrEnum):
+    FRESHMAN = 'FR', _('Freshman')
+    SOPHOMORE = 'SO', _('Sophomore')
+    JUNIOR = 'JR', _('Junior')
+    SENIOR = 'SR', _('Senior')
+    GRADUATE = 'GR', _('Graduate')
+
+
+class Vehicle(models.ChoiceIntEnum):
+    CAR = 1, 'Carriage'
+    TRUCK = 2
+    JET_SKI = 3
+
+    __empty__ = _('(Unknown)')
+
+
+class Gender(models.ChoiceStrEnum):
+    MALE = 'M'
+    FEMALE = 'F'
+    NOT_SPECIFIED = 'X'
+
+    __empty__ = _('(Undeclared)')
+
+
+Medal = models.ChoiceStrEnum('Medal', 'GOLD SILVER BRONZE')
+Place = models.ChoiceIntEnum('Place', 'FIRST SECOND THIRD')
+
+
+class Separator(bytes, enum.Enum, metaclass=ChoiceEnumMeta):
```

Comment:
We added many tests with custom enums that use `metaclass=ChoiceEnumMeta` if we want to support this usage then documentation is missing (at least a short section with a single example, maybe `MoonLandings`).

Reply:
`ChoiceIntEnum` and `ChoiceStrEnum` were explicitly defined to avoid the more complex signature, but they are not compulsory. Enums can be subclassed unless they have members defined in them. We could have skipped defining them (although for `str` we wanted to define `_generate_next_value_()` to make the Functional API work sanely out of the box).

Of the various tested examples, are there any data types that we want to provide an implementation for like `ChoiceIntEnum` and `ChoiceStrEnum`? Also do we like those names, or would we prefer `IntegerChoiceEnum` and `TextChoiceEnum` or even just `IntegerChoices` and `CharChoices` / `TextChoices` as this would tie up nicely with the fields they are intended for?

Reply:
I don't think we should include other choice enums out of the box -- but maybe that's just me. On the other side, if possible, I'd rather people not have to use `ChoiceEnumMeta` explicitly -- I prefer 
```python
class Separator(bytes, ChoiceEnum):
```
if it can be arranged.

With respect to the names, I took the cue from `enum.IntEnum`, thinking that this is a cute hack on enums rather than a major feature for fields. If we see it as an important Django feature, perhaps the naming you suggest is preferable. No strong opinion either way.

Reply:
Ok. So I agree that we want to make this as simple for people as possible, and to that end after some more thought I think that the names should also change. After all, these custom enum implementations are designed to be used for the specific purpose of working with `.choices`.

```python
class ChoicesMeta(enum.EnumMeta):
    ...


class Choices(enum.Enum, metaclass=ChoicesMeta):
    pass


class IntegerChoices(int, Choices):
    ...


class TextChoices(str, Choices):
    ...
```
So usage would then be quite descriptive:
```python
class Deck(models.Model):
    class Suit(models.IntegerChoices):
        DIAMOND = 1
        SPADE = 2
        HEART = 3
        CLUB = 4

    suit = models.PositiveIntegerField(choices=Suit.choices)
    ...


class Event(models.Model):
    Medal = models.TextChoices('Medal', 'GOLD SILVER BRONZE')

    name = models.CharField(max_length=100)
    date = models.DateField()
    medal = models.CharField(choices=Medal.choices)
    ...
```

And this would also make defining new choices very simple:
```python
class Separator(bytes, models.Choices):
    ...
```

Reply:
Works for me.